File: src/main/java/org/svomz/apps/finances/ports/adapters/persistence/mongo/MongoAccountRepository.java
package org.svomz.apps.finances.ports.adapters.persistence.mongo;

import static com.mongodb.client.model.Filters.*;

import com.google.common.base.Preconditions;

import com.mongodb.MongoClient;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.MongoDatabase;

import org.bson.Document;
import org.springframework.stereotype.Component;
import org.svomz.apps.finances.domain.model.Account;
import org.svomz.apps.finances.domain.model.AccountId;
import org.svomz.apps.finances.domain.model.AccountRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import javax.inject.Inject;

/**
 * Created by eric on 18/04/17.
 */
@Component
public class MongoAccountRepository implements AccountRepository {

  private final MongoClient mongoClient;
  private final MongoDatabase mongoDatabase;

  @Inject
  public MongoAccountRepository(MongoClient mongoClient) {
    this.mongoClient = Preconditions.checkNotNull(mongoClient);
    this.mongoDatabase = this.mongoClient.getDatabase("finances");
  }

  @Override
  public List<Account> findAll() {
    MongoCursor<Document> cursor = this.mongoDatabase.getCollection("accounts")
      .find().iterator();

    List<Account> results = new ArrayList<>();
    while (cursor.hasNext()) {
      results.add(this.map(cursor.next()));
    }

    cursor.close();
    return results;
  }

  @Override
  public Optional<Account> find(String id) {
    Document first = this.mongoDatabase.getCollection("accounts")
      .find(eq("accountId", id)).first();

    if (null == first) {
      return Optional.empty();
    }

    return Optional.of(this.map(first));
  }

  @Override
  public AccountId nextIdentity() {
    return new AccountId(UUID.randomUUID().toString());
  }

  @Override
  public Account create(Account account) {
    this.mongoDatabase.getCollection("accounts")
      .insertOne(this.map(account));

    return account;
  }

  @Override
  public void update(Account account) {
    this.mongoDatabase.getCollection("accounts")
      .updateOne(eq("accountId", account.getAccountId().getId()), new Document("$set", this.map(account)));
  }

  private Account map(Document next) {
    return new Account(
      new AccountId(next.getString("accountId")),
      next.getString("description")
    );
  }

  private Document map(Account account) {
    return new Document("accountId", account.getAccountId().getId())
      .append("description", account.getDescription());
  }
}

File: web/app/services/port/adapter/secondary/persistence/MongoAccountRepository.scala
package services.port.adapter.secondary.persistence

import java.util.UUID

import com.google.inject.Inject
import org.mongodb.scala.MongoDatabase
import org.mongodb.scala.bson.collection.immutable.Document
import org.mongodb.scala.model.Filters._
import org.mongodb.scala.model.Updates._
import org.svomz.apps.finances.core.domain.model.{Account, AccountId, AccountRepository}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

class MongoAccountRepository @Inject() (db: MongoDatabase) extends AccountRepository {

  private val collection = db.getCollection("accounts")

  override def update(account: Account): Future[Account] = {
    collection.updateOne(equal("accountId", account.id.value), set("description", account.name)).toFuture() map { _ => account }
  }

  override def delete(account: Account): Future[Account] = {
    collection.deleteOne(equal("accountId", account.id.value)).toFuture() map { _ => account }
  }

  override def create(account: Account): Future[Account] = {
    collection.insertOne(toDocument(account)).toFuture() map { _ => account }
  }

  override def fetch(no: AccountId): Future[Option[Account]] = {
    collection.find(equal("accountId", no.value)).first().toFuture() map { s =>
      if (s.isEmpty) None else Some(fromDocument(s.head))
    }
  }

  override def all: Future[List[Account]] = {
    collection.find().toFuture() map { s =>
      s map { d => fromDocument(d)} toList
    }
  }

  override def nextIdentity: Future[AccountId] = Future { AccountId(UUID.randomUUID().toString) }

  private def toDocument(account: Account): Document = {
    Document("accountId" -> account.id.value, "description" -> account.name)
  }

  def fromDocument(doc: Document): Account = {
    Account(AccountId(doc("accountId").asString().getValue), doc("description").asString().getValue)
  }
}
